### PR TITLE
chore: add progressive rollout

### PR DESCRIPTION
**Pull request by ono-max**
Implementation based on https://github.com/bucketeer-io/bucketeer/pull/282.

https://github.com/bucketeer-io/bucketeer/issues/308

### CODE REVIEW

File: pkg/feature/api/feature.go [hunk 1]
```diff
@@ -1215,6 +1275,27 @@ func (s *FeatureService) UpdateFeatureVariations(
 		}
 		return nil, dt.Err()
 	}
+	runningProgressiveRolloutExists, err := s.existsRunningProgressiveRollout(ctx, req.Id, req.EnvironmentNamespace)
```
Comment (masaaania) [inline pkg/feature/api/feature.go:1278]: question:

We need to validate the existence of running progressive rollout on other function like ArchiveFeature(), CloneFeature() and so on?

File: pkg/autoops/api/error.go [hunk 2]
```diff
@@ -95,8 +95,80 @@ var (
...
+		codes.InvalidArgument,
+		"autoops progressive rollout: clause variation id must be specified",
+	)
+	statusProgressiveRolloutClauseSchedulesRequired = gstatus.New(
+		codes.InvalidArgument,
+		"autoops progressive rollout: at least one clause schedule must be specified",
+	)
+	statusProgressiveRolloutClauseInvalidIncrements = gstatus.New(
+		codes.InvalidArgument,
+		"autoops progressive rollout: clause increments is invalid.",
```
Comment (masaaania) [inline pkg/autoops/api/error.go:128]: nit:

```suggestion
		"autoops progressive rollout: clause increments is invalid",
```

File: pkg/autoops/api/error.go [hunk 3]
```diff
@@ -95,8 +95,80 @@ var (
...
+	)
+	statusProgressiveRolloutClauseUnknownInterval = gstatus.New(
+		codes.InvalidArgument,
+		"autoops progressive rollout: clause interval is unknown.",
```
Comment (masaaania) [inline pkg/autoops/api/error.go:132]: nit:

```suggestion
		"autoops progressive rollout: clause interval is unknown",
```

File: pkg/autoops/api/error.go [hunk 4]
```diff
@@ -95,8 +95,80 @@ var (
...
+	)
+	statusProgressiveRolloutInvalidVariationSize = gstatus.New(
+		codes.FailedPrecondition,
+		"autoops progressive rollout: the number of variations must be equal to 2",
+	)
+	statusProgressiveRolloutInvalidScheduleSpans = gstatus.New(
+		codes.FailedPrecondition,
+		"autoops progressive rollout: The span of time for each scheduled time must be at least 5 minutes.",
```
Comment (masaaania) [inline pkg/autoops/api/error.go:140]: nit:

```suggestion
		"autoops progressive rollout: the span of time for each scheduled time must be at least 5 minutes",
```

**Comment by masaaania** [comment]
It would be helpful to separate large PRs into smaller PRs next time 🙏

**Comment by ono-max** [comment]
@masaaania 

Thank you for your review, and sorry for bothering reviewing because of big changes 🙏

File: pkg/feature/api/feature.go (hunk 1 above)
Comment (ono-max) [reply pkg/feature/api/feature.go:1278]: Talked at discord.

File: pkg/autoops/domain/progressive_rollout.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+		Clause:    nil,
+		CreatedAt: now,
+		UpdatedAt: now,
+	}}
+	if manual != nil {
+		if err := p.addManualScheduleClause(manual); err != nil {
+			return nil, err
+		}
+	}
+	if template != nil {
```
[the remark this answers is not in the document]
Comment (ono-max) [reply pkg/autoops/domain/progressive_rollout.go:62]: > The current condition could override the Clause if both are set by mistake.

I think that we need to check this case when validating requests. Let me change it. Thank you!

File: pkg/autoops/api/progressive_rollout_operation.go [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+			Weight:    schedule.Weight,
+		},
+		{
+			Variation: anotherVID,
+			Weight:    totalVariationWeight - schedule.Weight,
+		},
+	}, nil
+}
+
+func findAnotherVariationID(
```
[the remark this answers is not in the document]
Comment (ono-max) [reply pkg/autoops/api/progressive_rollout_operation.go:245]: Talked at disocrd.

File: pkg/autoops/api/error.go [hunk 7]
```diff
@@ -95,8 +95,112 @@ var (
...
+	)
+	statusProgressiveRolloutFeatureIDRequired = gstatus.New(
+		codes.InvalidArgument,
+		"autoops progressive rollout: feature id must be specified",
+	)
+	statusProgressiveRolloutClauseRequired = gstatus.New(
+		codes.InvalidArgument,
+		"autoops progressive rollout: at least one clause must be specified",
+	)
+	statusIncorrecctProgressiveRolloutClause = gstatus.New(
```
Comment (masaaania) [inline pkg/autoops/api/error.go:114]: nit:

```suggestion
	statusIncorrectProgressiveRolloutClause = gstatus.New(
```

File: pkg/autoops/api/progressive_rollout.go [hunk 8]
```diff
@@ -0,0 +1,1011 @@
...
+)
+
+const (
+	fiveMinutes     = 5 * time.Minute
+	listRequestSize = 500
+)
+
+var (
+	errProgressiveRolloutAutoOpsWebhookClauseExists  = errors.New("progressive rollout: ")
+	errProgressiveRolloutAutoOpsDatetimeClauseExists = errors.New("")
```
Comment (masaaania) [inline pkg/autoops/api/progressive_rollout.go:46]: errors.New()'s args is empty.

**Comment by ono-max** [comment]
@masaaania 

@cre8ivejp 

Could you take a look at this PR? Thank you!

**Review by masaaania — DISMISSED** [review]
LGTM!

File: pkg/autoops/api/error.go [hunk 9]
```diff
@@ -143,6 +143,10 @@ var (
 		codes.FailedPrecondition,
 		"autoops progressive rollout: can not use progressive rollout when feature is disabled",
 	)
+	statusProgressiveRolloutFeaturePrerequisiteExistss = gstatus.New(
+		codes.FailedPrecondition,
+		"autoops progressive rollout: can not use progressive rollout when feature has prerequisites",
```
Comment (cre8ivejp) [inline pkg/autoops/api/error.go:148]: nit: `autoops: can not create a progressive rollout when the feature has prerequisites`

File: pkg/locale/localizedata/ja.yaml [hunk 10]
```diff
@@ -29,6 +29,7 @@ IncompatibleOpsType: "対象のオペレーションタイプに対応してい
 InvalidVariationSize: "featureのvariationsの数が不正です"
 InvalidScheduleSpans: "スケジュールされた時間の間隔は少なくとも5分以上必要です"
 FeatureDisabled: "featureが有効でありません"
+FeaturePrerequisiteExists: "featureにprerequisiteがあります"
```
Comment (cre8ivejp) [inline pkg/locale/localizedata/ja.yaml:32]: This message should be more specific, like we are in the logs.

File: pkg/locale/localizer.go [hunk 11]
```diff
@@ -54,6 +54,7 @@ const (
 	InvalidScheduleSpans                     = "InvalidScheduleSpans"
 	WaitingOrRunningProgressiveRolloutExists = "WaitingOrRunningProgressiveRolloutExists"
 	FeatureDisabled                          = "FeatureDisabled"
+	FeaturePrerequisiteExists                = "FeaturePrerequisiteExists"
```
Comment (cre8ivejp) [inline pkg/locale/localizer.go:57]: nit `FeatureHasPrerequisites`

File: pkg/autoops/api/error.go [hunk 12]
```diff
@@ -147,6 +147,10 @@ var (
 		codes.FailedPrecondition,
 		"autoops progressive rollout: can not use progressive rollout when feature has prerequisites",
 	)
+	statusProgressiveRolloutFeatureTargetExists = gstatus.New(
+		codes.FailedPrecondition,
+		"autoops progressive rollout: can not use progressive rollout when feature has targets",
```
Comment (cre8ivejp) [inline pkg/autoops/api/error.go:152]: nit: `autoops: can not create a progressive rollout when the feature flag has individual targeting set`

File: pkg/locale/localizedata/ja.yaml [hunk 13]
```diff
@@ -30,6 +30,7 @@ InvalidVariationSize: "featureのvariationsの数が不正です"
 InvalidScheduleSpans: "スケジュールされた時間の間隔は少なくとも5分以上必要です"
 FeatureDisabled: "featureが有効でありません"
 FeaturePrerequisiteExists: "featureにprerequisiteがあります"
+FeatureTargetExists: "featureにtargetがあります"
```
Comment (cre8ivejp) [inline pkg/locale/localizedata/ja.yaml:33]: This message should be more specific, like we are in the logs.

File: pkg/locale/localizer.go [hunk 14]
```diff
@@ -55,6 +55,7 @@ const (
 	WaitingOrRunningProgressiveRolloutExists = "WaitingOrRunningProgressiveRolloutExists"
 	FeatureDisabled                          = "FeatureDisabled"
 	FeaturePrerequisiteExists                = "FeaturePrerequisiteExists"
+	FeatureTargetExists                      = "FeatureTargetExists"
```
Comment (cre8ivejp) [inline pkg/locale/localizer.go:58]: nit: `FeatureHasIndividualTargeting`

File: pkg/autoops/api/error.go [hunk 15]
```diff
@@ -151,6 +151,10 @@ var (
 		codes.FailedPrecondition,
 		"autoops progressive rollout: can not use progressive rollout when feature has targets",
 	)
+	statusProgressiveRolloutFeatureRuleExists = gstatus.New(
+		codes.FailedPrecondition,
+		"autoops progressive rollout: can not use progressive rollout when feature has rules",
```
Comment (cre8ivejp) [inline pkg/autoops/api/error.go:156]: nit: `autoops: can not create a progressive rollout when the feature flag has targeting rules set`

File: pkg/locale/localizedata/ja.yaml [hunk 16]
```diff
@@ -31,6 +31,7 @@ InvalidScheduleSpans: "スケジュールされた時間の間隔は少なくと
 FeatureDisabled: "featureが有効でありません"
 FeaturePrerequisiteExists: "featureにprerequisiteがあります"
 FeatureTargetExists: "featureにtargetがあります"
+FeatureRuleExists: "featureにruleがあります"
```
Comment (cre8ivejp) [inline pkg/locale/localizedata/ja.yaml:34]: This message should be more specific, like we are in the logs.

File: pkg/locale/localizer.go [hunk 17]
```diff
@@ -56,6 +56,7 @@ const (
 	FeatureDisabled                          = "FeatureDisabled"
 	FeaturePrerequisiteExists                = "FeaturePrerequisiteExists"
 	FeatureTargetExists                      = "FeatureTargetExists"
+	FeatureRuleExists                        = "FeatureRuleExists"
```
Comment (cre8ivejp) [inline pkg/locale/localizer.go:59]: nit: `FeatureHasRules`

File: pkg/locale/localizer.go [hunk 18]
```diff
@@ -57,6 +57,8 @@ const (
 	FeaturePrerequisiteExists                = "FeaturePrerequisiteExists"
 	FeatureTargetExists                      = "FeatureTargetExists"
 	FeatureRuleExists                        = "FeatureRuleExists"
+	AutoOpsDatetimeClauseExists              = "AutoOpsDatetimeClauseExists"
```
Comment (cre8ivejp) [inline pkg/locale/localizer.go:60]: You can shorten it by renaming it `AutoOpsHasDatetime` or `AutoOpsHasDatetimeSet`.
Same for the other var names.

File: pkg/autoops/api/progressive_rollout.go [hunk 19]
```diff
@@ -65,6 +76,13 @@ func (s *AutoOpsService) CreateProgressiveRollout(
 		return nil, dt.Err()
 	}
 	err = s.mysqlClient.RunInTransaction(ctx, tx, func() error {
+		// There are two reasons of validating auto ops rules:
+		// 1. We have to remove some test cases in `TestCreateProgressiveRollout`.
```
Comment (cre8ivejp) [inline pkg/autoops/api/progressive_rollout.go]: I didn't understand the connection between removing the `TestCreateProgressiveRollout` and validating it inside the transaction.

File: pkg/autoops/api/progressive_rollout.go [hunk 20]
```diff
@@ -42,8 +42,8 @@ const (
 )
 
 var (
-	errProgressiveRolloutAutoOpsWebhookClauseExists  = errors.New("progressive rollout: ")
-	errProgressiveRolloutAutoOpsDatetimeClauseExists = errors.New("")
+	errProgressiveRolloutAutoOpsWebhookClauseExists  = errors.New("progressive rollout: can not use progressive rollout when auto ops date time clause exists")
+	errProgressiveRolloutAutoOpsDatetimeClauseExists = errors.New("progressive rollout: can not use progressive rollout when auto ops webhook clause exists")
```
Comment (cre8ivejp) [inline pkg/autoops/api/progressive_rollout.go]: nit: `autoops: can not create a progressive rollout when the webhook is set in the auto ops`

File: pkg/autoops/api/progressive_rollout.go [hunk 21]
```diff
@@ -42,8 +42,8 @@ const (
 )
 
 var (
-	errProgressiveRolloutAutoOpsWebhookClauseExists  = errors.New("progressive rollout: ")
-	errProgressiveRolloutAutoOpsDatetimeClauseExists = errors.New("")
+	errProgressiveRolloutAutoOpsWebhookClauseExists  = errors.New("progressive rollout: can not use progressive rollout when auto ops date time clause exists")
```
Comment (cre8ivejp) [inline pkg/autoops/api/progressive_rollout.go]: nit: `autoops: can not create a progressive rollout when the schedule is set in the auto ops`

**Comment by ono-max** [comment]
@cre8ivejp 

Updated 👍

**Comment by cre8ivejp** [comment]
@ono-max 
Looks good! Thank you for the amazing work!
I will approve this PR later!

**Comment by ono-max** [comment]
Rebased

**Comment by ono-max** [comment]
Rebased

**Comment by ono-max** [comment]
Rebased

**Comment by cre8ivejp** [comment]
Closing in favor of https://github.com/bucketeer-io/bucketeer/pull/550

**Pull request closed without merging**